src/Microsoft.Data.SqlClient/src/Microsoft/Data/ProviderBase/DbReferenceCollection.cs

```diff
@@ -0,0 +1,279 @@
+                                    NotifyItem(message, _items[index].RefInfo, value);
+                                    _items[index].RemoveTarget();
+                                }
+                                Debug.Assert(!_items[index].TryGetTarget(out object _), "Unexpected target after notifying");
+                            }
+                            _estimatedCount = 0;
+                        }
+
+                        // Shrink collection (if needed)
+                        if (_items.Length > 100)
+                        {
+                            _lastItemIndex = 0;
+                            _items = new CollectionEntry[DefaultCollectionSize];
+                        }
+                    }
+                    finally
+                    {
+                        _isNotifying = false;
+                    }
+                }
+            }
+            finally
+            {
+                ExitItemLockIfNeeded(lockObtained);
+            }
+        }
+
+        abstract protected void NotifyItem(int message, int refInfo, object value);
+
+        abstract public void Remove(object value);
+
+        protected void RemoveItem(object value)
+        {
+            Debug.Assert(null != value, "RemoveItem with null");
+
+            bool lockObtained = false;
+            try
+            {
+                TryEnterItemLock(ref lockObtained);
+
+                if (lockObtained)
+                {
+                    // Find the value, and then remove the target from our collection
+                    if (_estimatedCount > 0)
+                    {
+                        for (int index = 0; index <= _lastItemIndex; ++index)
+                        {
+                            if (_items[index].TryGetTarget(out object target) && value == target)
+                            {
+                                _items[index].RemoveTarget();
+                                _estimatedCount--;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                ExitItemLockIfNeeded(lockObtained);
+            }
+        }
+
+        // This is polling lock that will abandon getting the lock if _isNotifying is set to true
+        private void TryEnterItemLock(ref bool lockObtained)
+        {
+            // Assume that we couldn't take the lock
+            lockObtained = false;
+            // Keep trying to take the lock until either we've taken it, or the collection is being notified
+            while ((!_isNotifying) && (!lockObtained))
+            {
+                Monitor.TryEnter(_itemLock, LockPollTime, ref lockObtained);
+            }
+        }
+
+        private void ExitItemLockIfNeeded(bool lockObtained)
+        {
+            if (lockObtained)
+            {
+                Monitor.Exit(_itemLock);
+            }
+        }
+    }
+}
```

Comment:
why this change is added here? another PR has merged to address this.

Reply:
Because I could not get it merged in for some reason.